src/intersection-observer-polyfill.js

```diff
@@ -0,0 +1,270 @@
+ *   top: number,
+ *   bottom: number,
+ *   left: number,
+ *   right: number,
+ *   width: number,
+ *   height: number,
+ *   x: number,
+ *   y: number,
+ * }}
+ */
+export let DOMRect;
+
+/**
+ * The IntersectionObserverPolyfill class lets any element receive its
+ * intersection data with the viewport. It acts like native browser supported
+ * IntersectionObserver.
+ * The IntersectionObserver receives a callback function and an optional option
+ * as params. Whenever the element intersection ratio cross a threshold value,
+ * IntersectionObserverPolyfill will call the provided callback function with
+ * the change entry.
+ */
+export class IntersectionObserverPolyfill {
+  /**
+   * @param {!function()} callback.
+   * @param {Object=} opt_option
+   */
+  constructor(callback, opt_option) {
+    /** @private @const {function()} */
+    this.callback_ = callback;
+
+    /**
+     * A list of threshold, sorted in increasing numeric order
+     * @private @const {!Array}
+     */
+    this.threshold_ = opt_option ? opt_option.threshold : [0];
+    this.threshold_ = this.threshold_.sort();
+    dev().assert(this.threshold_[0] >= 0 &&
+        this.threshold_[this.threshold_.length - 1] <= 1,
+        'Threshold should only contain value range from 0 to 1');
+
+    /** @private {?Element} */
+    this.element_ = null;
+
+    /** @private {boolean} */
+    this.isAmpElement_ = false;
+
+    /**
+     * A cached value to store getBoundingClientRect() value
+     * if element is not an AMP element
+     * @private {?DOMRect}
+     */
+    this.elementRectForNonAMP_ = null;
+
+    /**
+     * The prev threshold slot which the previous ratio fills
+     * Range [0, this.threshold_.length]
+     * TODO: Do we always want to call callback at the first tick?
+     * @private {number}
+     */
+    this.prevThresholdSlot_ = 0;
+
+    // TODO: create the PositionObserver class, or create listener for position
+    // info if inside an iframe
+
+    // TODO: Add the postMessageApi logic outside IntersectionObserverPolyfill.
+  }
+
+  /**
+   * Provide a way to observer the intersection change for a specific element
+   * Please note that the IntersectionObserverPolyfill only allow to observe one
+   * element at a time.
+   * TODO: Support observing multiple elements.
+   * @param {!Element} element
+   */
+  observe(element) {
+    // Reset cached value for a new element
+    this.isAmpElement_ = false;
+    this.elementRectForNonAMP_ = null;
+    this.prevThresholdSlot_ = 0;
+    // TODO: Need a good way to check if an element is AMP element
+    if (element.isBuilt && element.isBuilt()) {
+      this.isAmpElement_ = true;
+    }
+    this.element_ = element;
+  }
+
+  /**
+   * Tick function that update the DOMRect of the root of observed element.
+   * Caller needs to make sure to pass in the correct container.
+   * @param {!./layout-rect.LayoutRectDef} viewport.
+   * @param {./layout-rect.LayoutRectDef=} opt_container. relative to viewport
```

Comment:
Well by using the word `container` I want to convey the idea that the element should be inside the container and know nothing about the outside window. As we make the assumption `#getLayout` returned value is the layout rect relative to container instead of the page. `offset` sounds like a pure offset value to me.